ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Monad.hs

```diff
@@ -0,0 +1,126 @@
+{-# LANGUAGE BangPatterns      #-}
+{-# LANGUAGE DeriveFoldable    #-}
+{-# LANGUAGE DeriveFunctor     #-}
+{-# LANGUAGE DeriveTraversable #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE Rank2Types        #-}
+{-# LANGUAGE TypeFamilies      #-}
+
+-- | A monad for the interface to the ledger layer
+module Ouroboros.Consensus.Ledger.Monad (
+    -- * Ledger events
+    AuxLedgerEvent
+  , VoidLedgerEvent
+    -- * Ledger monad
+  , LedgerResult (..)
+  , LedgerT (..)
+  , coerceLedgerT
+  , evalLedgerT
+  , hoistLedgerT
+  , ledgerT
+  , liftLedgerT
+  , runLedgerT
+  , withLedgerT
+  ) where
+
+import           Control.Monad.Except
+import           Control.Monad.Morph (MFunctor (..))
+import           Control.Monad.Reader (ReaderT (ReaderT), runReaderT,
+                     withReaderT)
+import           Control.Monad.State.Strict (StateT (StateT), runStateT)
+import           Data.Functor ((<&>))
+import           Data.Kind (Type)
+
+{-------------------------------------------------------------------------------
+  Ledger monad
+-------------------------------------------------------------------------------}
+
+-- | Event emitted by the ledger
+--
+-- TODO we call this 'AuxLedgerEvent' to differentiate from 'LedgerEvent' in
+-- 'InspectLedger'. When that module is rewritten to make use of ledger
+-- derived events, we may rename this type.
+type family AuxLedgerEvent l :: Type
+
+-- | A 'Data.Void.Void' isomorph for ledgers that have no events.
+--
+data VoidLedgerEvent
+
+-- | The monad transformer used when interacting with the ledger
+newtype LedgerT l m a = LedgerT {
+    unLedgerT :: forall ev. ReaderT (AuxLedgerEvent l -> ev) (StateT [ev] m) a
+  }
+
+-- | The primary way to build a 'LedgerT' computation
+ledgerT :: Functor m => m (LedgerResult l a) -> LedgerT l m a
+ledgerT m =
+    LedgerT $ ReaderT $ \r -> StateT $ \s ->
+      m <&> \case
+        LedgerResult {
+            lrEvents = events
+          , lrResult = !a
+          } -> (a, s <> map r events)
```

Comment:
TODO how strict do we want to be in this state? Our idea was to leave them as a thunk so that users who don't want them can simply never force them. Instead of needing to introduce a flag etc. But that will be one `<>` thunk per block, etc.

Reply:
I see, you are using laziness in place of the `EPDiscard` option.

Reply:
Yeah, that's the goal, at least.

Reply:
Edsko and I decided to drop the monad from the interface.